pkg/node/manager/manager.go

```diff
@@ -280,24 +280,22 @@ func (m *Manager) backgroundSync() {
 func (m *Manager) NodeUpdated(n node.Node) {
 	log.Debugf("Received node update event from %s: %#v", n.Source, n)
 	nodeIdentity := n.Identity()
-	var nodeIP, nodeIP4 net.IP
 	dpUpdate := true
+	nodeIP := n.GetNodeIP(false)
+
+	remoteHostIdentity := identity.ReservedIdentityHost
+	if option.Config.EnableRemoteNodeIdentity {
+		remoteHostIdentity = identity.ReservedIdentityRemoteNode
+	}
 
 	for _, address := range n.IPAddresses {
-		// Map the Cilium internal IP to the reachable node IP so it
-		// can be routed via the overlay. Routing via overlay is always
-		// done via public v4 address hence n.GetNodeIP(false).
-		if address.Type == addressing.NodeCiliumInternalIP {
-			nodeIP = n.GetNodeIP(false)
-			if address.IP.To4() != nil {
-				nodeIP4 = nodeIP
-			}
-		} else {
-			continue
+		var tunnelIP net.IP
+		if address.Type == addressing.NodeCiliumInternalIP || option.Config.EncryptNode {
+			tunnelIP = nodeIP
 		}
 
-		isOwning := ipcache.IPIdentityCache.Upsert(address.IP.String(), nodeIP, n.EncryptionKey, nil, ipcache.Identity{
-			ID:     identity.ReservedIdentityHost,
+		isOwning := ipcache.IPIdentityCache.Upsert(address.IP.String(), tunnelIP, n.EncryptionKey, nil, ipcache.Identity{
+			ID:     remoteHostIdentity,
```

Comment:
Does this path get hit for the local node as well? Is it possible we unintentionally end up updating the ipcache with local IPs pointing to ID: `identity.ReservedIdentityRemoteNode`?

Reply:
I have validated that the local host entries are correct:
```
fe80::a49e:ebff:fe11:91ef/128   1 0 0.0.0.0
10.0.2.15/32                    1 0 0.0.0.0
10.16.228.47/32                 1 0 0.0.0.0
f00d::a10:0:0:9a42/128          1 0 0.0.0.0
fc00::10ca:1/128                1 0 0.0.0.0
192.168.34.11/32                1 0 0.0.0.0
192.168.33.11/32                1 0 0.0.0.0
```

Although I think you are correct that it can be called and the reason why it works is because  `syncEndpointsAndHostIPs()` is called first and will insert the local IPs with the correct identity and because it uses `source.Local`, it can't be overwritten by any subsequent updates.

Reply:
I've changed the condition to:
```
if option.Config.EnableRemoteNodeIdentity && n.Source != source.Local {
```